apps/prairielearn/src/sprocs/ip_to_mode.sql

```diff
@@ -22,63 +20,66 @@ BEGIN
+            (ip_to_mode.date BETWEEN r.access_start AND access_end) AS reservation_active,
+            l.id AS location_id,
+            l.filter_networks AS location_filter_networks
         FROM
-            pt_sessions AS s
-            JOIN pt_locations AS loc ON (loc.id = s.location_id)
+            pt_reservations AS r
+            JOIN pt_enrollments AS e ON (e.id = r.enrollment_id)
+            JOIN pt_sessions AS s ON (s.id = r.session_id)
+            LEFT JOIN pt_locations AS l ON (l.id = s.location_id)
         WHERE
-            s.id = v_session_id;
-
-        IF FOUND THEN
-            -- The reservation is in a testing center location.
-            IF v_filter_networks THEN
-                -- The reservation is in a testing center location that requires
-                -- network filtering.
-
-                PERFORM *
-                FROM pt_location_networks AS ln
-                WHERE
-                    ln.location_id = v_location_id
-                    AND ip <<= ln.network;
-
-                IF FOUND THEN
-                    -- The user is physically inside the testing center. Set
-                    -- mode to 'Exam'.
-                    mode := 'Exam';
-                ELSE
-                    -- Although we have a checked-in reservation, the user is
-                    -- physically outside the testing center. Set mode to
-                    -- 'Public'.
-                    mode := 'Public';
-                END IF;
-            ELSE
-                -- The reservation is in a testing center location that doesn't
-                -- require network filtering. Set mode to 'Exam'.
+            e.user_id = ip_to_mode.authn_user_id
+            AND (
+                (r.access_end IS NULL and s.date BETWEEN ip_to_mode.date - '1 hour'::interval and ip_to_mode.date + '1 hour'::interval)
+                OR (ip_to_mode.date BETWEEN r.access_start AND r.access_end)
+            )
+    LOOP
+        IF reservation.location_id IS NULL THEN
+            -- The reservation isn't for a testing center location (it's a
+            -- course-run session). If the reservation is "active", we're
+            -- in 'Exam' mode, and we return immediately. Otherwise, we might
+            -- be in 'Public' mode, but we continue looping to see if we have
+            -- any other reservations that might put us in 'Exam' mode.
+            IF reservation.reservation_active THEN
                 mode := 'Exam';
+                RETURN;
             END IF;
-        ELSE
-            -- The reservation isn't for a testing center location (it's a
-            -- course-run session), so we set exam mode.
+
+            mode := 'Public';
+            CONTINUE;
+        END IF;
+
+        IF NOT reservation.location_filter_networks THEN
+            -- The reservation is in a testing center location that doesn't
+            -- require network filtering. Set mode to 'Exam'.
             mode := 'Exam';
+            RETURN;
```

Comment:
In this case I think we need to consider `reservation_active` just like we do for course sessions. Otherwise a student who has an exam in a non-IP-filtering center session will enter Exam mode 1 hour before the session starts, even though they are probably not yet in the testing center. This wouldn't be a security problem because they still couldn't access the actual exam, but it would be annoying for them if they were trying to do some last-minute studying.

Broadly speaking, I think a non-IP-filtering center session should be treated identically to a course session. Perhaps we could even combine these two code paths?

Reply:
This is a great point! To check my understanding, we'd end up with something like this?

```sql
        IF reservation.location_id IS NULL OR NOT reservation.location_filter_networks THEN
            IF reservation.reservation_active THEN
                mode := 'Exam';
                RETURN;
            END IF;

            mode := 'Public';
            CONTINUE;
        END IF;
```

Reply:
Yep, exactly!

Reply:
I made this change in 94ccbbd.